### PR TITLE
style: listviews closer to SIP-34

### PR DESCRIPTION
**Pull request by nytai**
### SUMMARY

- changes styles and pagination to match SIP-34. Datasets was the target but made a few changes to chartList and dashboardList since changes needed to be made to the underlying listview component. 
- Improves loading of the listview. The listview doesn't render until data is fetched and fetching new data uses a skeleton shimmer loading effect. 
- adds tooltips (dataset type, last modified, actions, owners)
- adds column option for fixed size.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Screen Shot 2020-06-18 at 3 23 21 PM
Screen Shot 2020-06-17 at 7 34 45 PM
Dashboards and Charts
Screen Shot 2020-06-17 at 5 43 33 PM
Screen Shot 2020-06-17 at 5 43 44 PM

### TEST PLAN

- toggle `ENABLE_REACT_CRUD_VIEWS` config flag, see changes in datasets/tables view

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Changes UI (behind config flag `ENABLE_REACT_CRUD_VIEWS`)
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/components/ListView/TableCollection.tsx
Comment (rusackas) [inline superset-frontend/src/components/ListView/TableCollection.tsx:34]: I see why you're adding `styled` here to apply props to the width, but there IS a possible performance implication here, if there wind up being a bajillion table cells, and Emotion starts bogging down. It may be more performant to use the prop to add a _class_ on the `td`, and then use emotion hire up in the DOM tree. I.e. 
```
const Table = styled.table`
  td{
    &.xs { width: 25px; }
    &.sm { width: 50px; }
    &.md { width: 75px; }
    &.lg { width: 100px; }
    &.xl { width: 150px; }
    &.xxl { width: 200px; }
  }
`
```

File: superset-frontend/src/components/Menu/SubMenu.tsx
Comment (rusackas) [inline superset-frontend/src/components/Menu/SubMenu.tsx:69]: children? ¯\\\_(ツ)_/¯

**Review by rusackas — CHANGES_REQUESTED** [review]
One potential Emotion performance pitfall, otherwise LGTM

File: superset-frontend/src/components/ListView/ListView.tsx
Comment (etr2460) [inline superset-frontend/src/components/ListView/ListView.tsx:129]: do you need this fragment here? in fact, it seems like some extra nesting with !useNewUIFilers, title, and filterable can be joined into a single statement

File: superset-frontend/src/components/ListView/ListView.tsx
Comment (etr2460) [inline superset-frontend/src/components/ListView/ListView.tsx:220]: these translation strings won't work at all. I think using string replacement is required instead, details are here: https://github.com/apache-superset/superset-ui/tree/master/packages/superset-ui-translation

although, now that i look closer, it seems like these are just indentation changes. so i guess it's up to you if you want to fix them or not. not sure how they snuck past in the first place :/ if it's not in scope to fix them here, let me know and i can clean up afterwards

File: superset-frontend/src/components/ListView/ListViewStyles.less
Comment (etr2460) [inline superset-frontend/src/components/ListView/ListViewStyles.less:56]: do we have a variable for this one?

File: superset-frontend/src/components/ListView/ListViewStyles.less
Comment (etr2460) [inline superset-frontend/src/components/ListView/ListViewStyles.less:77]: feels like we might have variables for these colors already, although i'm not sure. Maybe we should add them if not?

File: superset-frontend/src/components/ListView/TableCollection.tsx
Comment (etr2460) [inline superset-frontend/src/components/ListView/TableCollection.tsx:43]: this could be simpler if you have each if only return a number, then you append `px` on at the end. Note the final return would need to return 0 instead of empty string then.

on a side note, do you think a switch statement is better here? it would remove even more code duplication

File: superset-frontend/src/components/Pagination.tsx
Comment (etr2460) [inline superset-frontend/src/components/Pagination.tsx:37]: is using unicode a good idea for this, or should we add svg icons while we're in here making stuff closer to SIP-34?

File: superset-frontend/src/components/Pagination.tsx
Comment (etr2460) [inline superset-frontend/src/components/Pagination.tsx:67]: should this be `...` or `…` (3 periods vs. an ellipsis)?

File: superset-frontend/src/types/react-table-config.d.ts
Comment (etr2460) [inline superset-frontend/src/types/react-table-config.d.ts:121]: Let's define these strings as it's own `Size` type then reuse it in both places

File: superset-frontend/src/views/dashboardList/DashboardList.tsx
Comment (etr2460) [inline superset-frontend/src/views/dashboardList/DashboardList.tsx:558]: nit, `t` string here please!

File: superset-frontend/src/views/chartList/ChartList.tsx
Comment (etr2460) [inline superset-frontend/src/views/chartList/ChartList.tsx:555]: nit: `t` string here please!

File: superset-frontend/src/views/chartList/ChartList.tsx
Comment (etr2460) [inline superset-frontend/src/views/chartList/ChartList.tsx:523]: nit: `t` string here please!

File: superset-frontend/src/views/dashboardList/DashboardList.tsx
Comment (etr2460) [inline superset-frontend/src/views/dashboardList/DashboardList.tsx:514]: nit: `t` string here please!

File: superset-frontend/src/views/dashboardList/DashboardList.tsx
Comment (etr2460) [inline superset-frontend/src/views/dashboardList/DashboardList.tsx:529]: nit: `t` string here please!

File: superset-frontend/src/views/chartList/ChartList.tsx
Comment (etr2460) [inline superset-frontend/src/views/chartList/ChartList.tsx:546]: nit: `t` string here please!

File: superset-frontend/src/views/dashboardList/DashboardList.tsx
Comment (etr2460) [inline superset-frontend/src/views/dashboardList/DashboardList.tsx:513]: it seems like a lot of code is the same as in the above file, is there any way we can pull it out into its own component and reuse it?

File: superset-frontend/src/views/datasetList/DatasetList.tsx
Comment (etr2460) [inline superset-frontend/src/views/datasetList/DatasetList.tsx:218]: it would be nice to move these format strings into constants (either in this file or in a more global constant file) assuming that these date and time formats will be widely used in the future

File: superset-frontend/src/views/datasetList/DatasetList.tsx
Comment (etr2460) [inline superset-frontend/src/views/datasetList/DatasetList.tsx:541]: nit: `t` string here please!

File: superset-frontend/src/views/datasetList/DatasetList.tsx
Comment (etr2460) [inline superset-frontend/src/views/datasetList/DatasetList.tsx:550]: nit: `t` string here please!

**Review by etr2460 — COMMENTED** [review]
A bunch of comments, most of them pretty small.

I realized after making a lot of them that I commented a lot on code you didn't actually change (only made indentation changes to). If you don't want to address them, that's ok and I can take a second pass later on. Although if you're feeling up to it, i think they're all pretty reasonable and not too much effort. I leave it up to you!

**Comment by mistercrunch** [comment]
> The listview doesn't render until data is fetched and fetching new data uses a skeleton shimmer loading effect.

(probably outside the scope of this PR) we should discuss the ideal page states in minimizing flicker as the page load. To me I think the ideal flow would be:

1. show the page with a menu and list view shell first (with no data, but headers and layout in-place), I think that means we have to do that rendering in the `<HEAD/>` before the page is shown
1. and then show the data in the table as it comes (trigger the data fetch in the tail)

Assuming fixed width and a full page height, going from 1 to 2 would really just show the data in the existing cells, showing a total of 2 states.

File: superset-frontend/src/components/ListView/ListView.tsx
Comment (nytai) [reply superset-frontend/src/components/ListView/ListView.tsx:220]: @etr2460 I wrote this before I really understood how translation strings worked. Any idea how to get the html (`<strong>`) interspersed in the translation string?

File: superset-frontend/src/views/dashboardList/DashboardList.tsx
Comment (nytai) [reply superset-frontend/src/views/dashboardList/DashboardList.tsx:513]: this is where it gets tricky, the code is largely the same but there are a few differences that make it difficult to isolate to a single component. It also couples the code so that changes in one view could result in unexpected changes in another view. I am planning to revisit this question once some of the design on these listviews has settled and is not likely to change drastically.

**Comment by nytai** [comment]
@mistercrunch it would be nice to just render the skeleton and just have the data swap in when it arrives. In order to do this we need a good skeleton component that can be constructed to reasonably represent the table layout in question, or have dummy data that can take the shape of the data in question

File: superset-frontend/src/components/Pagination.tsx
Comment (nytai) [reply superset-frontend/src/components/Pagination.tsx:37]: the mocks we have for SIP-34 uses unicode in these components :)

File: superset-frontend/src/components/ListView/ListView.tsx
Comment (etr2460) [reply superset-frontend/src/components/ListView/ListView.tsx:220]: ah, it's not possible currently. I know Max was trying to figure out a way to hack around it, but I don't think we ever came up with a solution.

If the bold styling is necessary here, then i'd say remove the `t` strings entirely. If not, then it might be better to remove the styling for now to support better translations

File: superset-frontend/src/views/dashboardList/DashboardList.tsx
Comment (nytai) [reply superset-frontend/src/views/dashboardList/DashboardList.tsx:513]: Something that comes to mind as a flexible solution that would allow customization across the different views is a BaseListViewPage component that these could inherit from and override certain methods as needed. This would remove some of the code redundancies while allowing flexibility in overriding certain behavior. I generally try to avoid inheritance in JS when possible, but it seems like there's a case for it here. Thoughts? 

Another approach might be to move this into hooks, ie `usePermissions`, `useFetchListData`, etc. though I'm not convinced this will really remove that much code redundancy

File: superset-frontend/src/components/Menu/SubMenu.tsx
Comment (nytai) [reply superset-frontend/src/components/Menu/SubMenu.tsx:69]: this is a legacy term from the FAB menu. I'll update

Comment (nytai) [reply superset-frontend/src/components/Menu/SubMenu.tsx:69]: This is legacy terminology from FAB's menu structure and it's the same in the react menu component, https://github.com/preset-io/incubator-superset/blob/master/superset-frontend/src/components/Menu/MenuObject.jsx#L28
Also I think children will conflict with react's built in props

File: superset-frontend/src/components/ListView/ListView.tsx
Comment (nytai) [reply superset-frontend/src/components/ListView/ListView.tsx]: removed the `t` strings. I've also asked design to look into it with translation in mind.

File: superset-frontend/src/components/ListView/ListViewStyles.less
Comment (nytai) [reply superset-frontend/src/components/ListView/ListViewStyles.less:80]: we don't have them and I don't see a huge advantage to making them variables; they're just black and black at half opacity.

File: superset-frontend/src/components/AvatarIcon.tsx
Comment (etr2460) [inline superset-frontend/src/components/AvatarIcon.tsx:62]: iconSize is typed as a number, so i don't think you need to wrap it in `Number()` here

File: superset-frontend/src/types/react-table-config.d.ts
Comment (etr2460) [inline superset-frontend/src/types/react-table-config.d.ts:70]: This probably should be `ColumnSize`, especially since that's how it's referenced on line 134

**Review by etr2460 — APPROVED** [review]
lgtm after addressing a couple comments

File: superset-frontend/src/components/AvatarIcon.tsx
Comment (nytai) [reply superset-frontend/src/components/AvatarIcon.tsx:62]: Ah yes, looks like I had two competing approaches here.

File: superset-frontend/src/components/AvatarIcon.tsx
Comment (rusackas) [inline superset-frontend/src/components/AvatarIcon.tsx:23]: Just a comment that's out of scope for this PR, and a bit pedantic, but it might be nice to just call this component `Tooltip` since I suspect _most_ of our Superset components going forward will be wrappers for some react-bootstrap/antd/whateverLibrary component. I would just name the tool semantically, lest we start calling everything a Wrapper.

Comment (rusackas) [reply superset-frontend/src/components/AvatarIcon.tsx:23]: ... and if this was due to namespacing issues, I would just rename the internal import, e.g. `import { Tooltip as RB_Tooltip } from 'react-bootstrap';

File: superset-frontend/src/components/ListView/ListView.tsx
Comment (rusackas) [inline superset-frontend/src/components/ListView/ListView.tsx:207]: Curious why "showing" and "of" were un-translated.

File: superset-frontend/src/components/ListView/ListViewStyles.less
Comment (rusackas) [reply superset-frontend/src/components/ListView/ListViewStyles.less:80]: FWIW, the LESS variables have pre-baked colors and opacity stops, so you can do things like:
`background-color: fade(@success, @opacity-heavy);`

Those opacity stops haven't been migrated to the JS Theme for Emotion, yet. Emotion doesn't have the fancy color mixins, so we'd have to find the nifty JS way to accomplish the same goal.

File: superset-frontend/src/components/ListView/ListView.tsx
Comment (nytai) [reply superset-frontend/src/components/ListView/ListView.tsx:207]: this isn't a good way to translate strings. Not all languages follow the same structure as english, so just simply translating single words does not translate to proper sentence in another language.

File: superset-frontend/src/components/AvatarIcon.tsx
Comment (nytai) [reply superset-frontend/src/components/AvatarIcon.tsx:23]: The reason it's tooltipWrapper is because it just allows for any `tooltip` to be rendered with the correct triggers. `tooltip` is a prop of tooltipWrapper that defaults to using the bootstrap tooltip.

**Comment by mistercrunch** [comment]
fire

**Comment by john-bodley** [comment]
@nytai I believe this PR introduced a regression for the `/superset/welcome` page, i.e., per the attached screenshot the `showing 1-6 of 6` is floating outside the table.

Screen Shot 2020-06-26 at 3 09 45 PM

**Pull request merged**
